Libraries/Opc.Ua.PubSub/Uadp/UadpNetworkMessage.cs

```diff
@@ -0,0 +1,1009 @@
+        /// Encode security header 
+        /// </summary>
+        /// <param name="encoder"></param>
+        private void EncodeSecurityHeader(BinaryEncoder encoder)
+        {
+            if ((ExtendedFlags1 & ExtendedFlags1EncodingMask.Security) != 0)
+            {
+                encoder.WriteByte("SecurityFlags", (byte)SecurityFlags);
+
+                encoder.WriteUInt32("SecurityTokenId", SecurityTokenId);
+                encoder.WriteByte("NonceLength", NonceLength);
+                MessageNonce = new byte[NonceLength];
+                encoder.WriteByteArray("MessageNonce", MessageNonce);
+
+                if ((SecurityFlags & SecurityFlagsEncodingMask.SecurityFooter) != 0)
+                {
+                    encoder.WriteUInt16("SecurityFooterSize", SecurityFooterSize);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Encode payload
+        /// </summary>
+        /// <param name="encoder"></param>
+        private void EncodePayload(BinaryEncoder encoder)
+        {
+            int payloadStartPositionInStream = encoder.Position;
+            if (m_uadpDataSetMessages.Count > 1
+                && (NetworkMessageContentMask & UadpNetworkMessageContentMask.PayloadHeader) != 0)
+            {                
+                //skip 2 * dataset count for each dataset payload size 
+                encoder.Position = encoder.Position + 2 * m_uadpDataSetMessages.Count;               
+            }
+            //encode dataset message payload
+            foreach (UadpDataSetMessage uadpDataSetMessage in m_uadpDataSetMessages)
+            {
+                uadpDataSetMessage.Encode(encoder);
+            }
+
+            if (m_uadpDataSetMessages.Count > 1
+                && (NetworkMessageContentMask & UadpNetworkMessageContentMask.PayloadHeader) != 0)
+            {
+                int payloadEndPositionInStream = encoder.Position;
+                encoder.Position = payloadStartPositionInStream;
+                foreach (UadpDataSetMessage uadpDataSetMessage in m_uadpDataSetMessages)
+                {                   
+                    encoder.WriteUInt16("Size", uadpDataSetMessage.PayloadSizeInStream);
+                }
+                encoder.Position = payloadEndPositionInStream;
+            }
+        }
+
+        /// <summary>
+        /// Encode security footer
+        /// </summary>
+        /// <param name="encoder"></param>
+        private void EncodeSecurityFooter(BinaryEncoder encoder)
+        {
+            if ((SecurityFlags & SecurityFlagsEncodingMask.SecurityFooter) != 0)
+            {
+               encoder.WriteByteArray("SecurityFooter", SecurityFooter);
+            }
+        }
+
+        /// <summary>
+        /// Encode signature
+        /// </summary>
+        /// <param name="encoder"></param>
+        private void EncodeSignature(BinaryEncoder encoder)
+        {
+           // encoder.WriteByteArray("Signature", Signature);
```

Comment:
is this a security signature?

Reply:
Yes, but it is not used since there is no security support yet.